 INDENG242 Homework 1

# Question 1

## (a)
We would expect that the RSS calculated using cubic regression tends to be smaller. That is because the cubic regression form has more degrees of freedom to fit the model with training set, resulting in a smaller RSS

## (b)
Since we already know that the true model is linear, we would expect that linear model does better job thatn the cubic model in terms of the test data.
    
## (c)
In terms of the training set, the cubic model will always do a better job than the linear model, meaning that having a lower RSS than the linear model does. Because it has more degrees of freedom to minimize the RSS

## (d)
Actually, it depends. In terms of the test sets, it depends on what the real underlying model is, like how non-linear it is. And it also depends on how much test data we have.  
If the underlying model is much or very non-linear, then we would expect the cubic regression model does a better job as explained before (having more degrees of freedom to fit the model).  
If we're given very little data, for example, only 4-5 sets of data, then calculating RSS for a cubic model is somehow non-sense because it has great variance. Then cubic model may be overfitting, resulting in a ridiculous fit.

# Question 2

With  $$\hat{\beta} = \frac{\sum_{i=1}^{n}{x_iy_i}}{\sum_{j=1}^{n}{x_j^{2}}} $$

we will have $$\hat{y_i} = x_i \frac{\sum_{i=1}^{n}{x_iy_i}}{\sum_{j=1}^{n}{x_j^{2}}} = {\sum_{k=1}^{n}{\frac{x_ix_k}{\sum_{j=1}^{n}{x_j^{2}}}}y_k}$$

And we had $$\hat{y_i} = {\sum_{i'=1}^{n}{a_{i'}y_{i'}}}$$

Therefore $$a_{i'} = \frac{x_ix_{i'}}{\sum_{j=1}^{n}{x_j^{2}}}$$

# Question 3

## (a)

In [1]:
import pandas as pd
import numpy as np

In [2]:
rogue = pd.read_csv('Rogue-242-Fall2021.csv')
rogue.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162 entries, 0 to 161
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   MonthNumeric  162 non-null    int64  
 1   MonthFactor   162 non-null    object 
 2   Year          162 non-null    int64  
 3   RogueSales    162 non-null    int64  
 4   Unemployment  162 non-null    float64
 5   RogueQueries  162 non-null    int64  
 6   CPIAll        162 non-null    float64
 7   CPIEnergy     162 non-null    float64
dtypes: float64(3), int64(4), object(1)
memory usage: 10.2+ KB


In [3]:
rogue_train = rogue[rogue['Year'] <= 2015]
rogue_testA = rogue[(rogue['Year'] >= 2016) & (rogue['Year'] <= 2019) ]
rogue_testB = rogue[(rogue['Year'] >= 2020) & (rogue['Year'] <= 2021) ]

In [4]:
import statsmodels.formula.api as smf

model = smf.ols(formula='RogueSales ~ RogueQueries + CPIEnergy + CPIAll + Unemployment',data=rogue_train).fit()

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.796
Model:                            OLS   Adj. R-squared:                  0.787
Method:                 Least Squares   F-statistic:                     88.71
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           1.47e-30
Time:                        12:52:24   Log-Likelihood:                -898.41
No. Observations:                  96   AIC:                             1807.
Df Residuals:                      91   BIC:                             1820.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept    -6.918e+04   1.42e+04     -4.884   

In [5]:
# Calculate Variance Inflation Factor for each explanatory variable

from statsmodels.stats.outliers_influence import variance_inflation_factor
import statsmodels.api as sm

def VIF(df, columns):
    
    values = sm.add_constant(df[columns]).values  # the dataframe passed to VIF must include the intercept term. We add it the same way we did before.
    num_columns = len(columns)+1
    vif = [variance_inflation_factor(values, i) for i in range(num_columns)]
    
    return pd.Series(vif[1:], index=columns)

In [6]:
# We remove Unemployment since it has a large p-value
model2 = smf.ols(formula='RogueSales ~ RogueQueries + CPIEnergy + CPIAll',data=rogue_train).fit()

print(model2.summary())

VIF(rogue_train,['RogueQueries','CPIEnergy','CPIAll'])

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     119.1
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           1.45e-31
Time:                        12:52:25   Log-Likelihood:                -898.57
No. Observations:                  96   AIC:                             1805.
Df Residuals:                      92   BIC:                             1815.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept    -6.908e+04   1.41e+04     -4.895   

RogueQueries    4.477880
CPIEnergy       1.471781
CPIAll          5.098905
dtype: float64

In [7]:
# We remove the CPIAll because it has a large VIF value
model3 = smf.ols(formula='RogueSales ~ RogueQueries + CPIEnergy',data=rogue_train).fit()

print(model3.summary())

VIF(rogue_train,['RogueQueries','CPIEnergy'])

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.731
Model:                            OLS   Adj. R-squared:                  0.725
Method:                 Least Squares   F-statistic:                     126.3
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           3.12e-27
Time:                        12:52:26   Log-Likelihood:                -911.69
No. Observations:                  96   AIC:                             1829.
Df Residuals:                      93   BIC:                             1837.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept     5168.8436   3301.430      1.566   

RogueQueries    1.003797
CPIEnergy       1.003797
dtype: float64

In [8]:
# We remove the CPIEnergy because it has a large p-value
model4 = smf.ols(formula='RogueSales ~ RogueQueries',data=rogue_train).fit()

print(model4.summary())

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.726
Model:                            OLS   Adj. R-squared:                  0.723
Method:                 Least Squares   F-statistic:                     249.5
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           3.39e-28
Time:                        12:52:29   Log-Likelihood:                -912.48
No. Observations:                  96   AIC:                             1829.
Df Residuals:                      94   BIC:                             1834.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept     1168.7582    777.053      1.504   

### i)
If we denote y to be the RogueSales and x to be the RogueQueries. The linear regression equation is 
$$y = 419.7103x+1168.7582 $$
    
How to interpret the coefficients: It means if the approximation of the number of Google search for "Rogue" increases by 1 unit, the Rogue Sales is estimated to increase by 419.7103 unit. Even if there is no search on the Google about it, the intercept stands for the mean value for the Rogue Sales.
    
### ii)
How to select the variable:  
Firstly we are given four variables which may leave an effect on the Rogue Sales. After taken all these four variables into OLS regression analysis, we found that the "Unemployment" has a quite large p-value(0.573). We then remove it from our model.  
After that, we use the VIF function to determine the multicollinearity. We found that the VIF value for CPIAll is vey large(>5). We therefore remove "CPIAll" too.  
And we are left with "CPIEnergy" and "RogueQueries". We still use the OLS Regression Analysis Result. Again the p-value for "CPIEnergy" is 0.216, which is also quite big. We remove the "CPIEnergy" too. 
Finally, we have "RogueQueries" left in our model.

### iii)
The coefficient of "RogueQueries" is positive, meaning that there will be an increase in Sales if the number of Google search increases. It seems quite reasonable. If people show interest in Rogue, they may search it online for the feedback or price, then people who are really interested and determined are likely to buy them, leading to an increase in the sales.
    
### iv)
According to the last OLS regression results, the R-Squared value is 0.726 which can be seen as a good enough sign of data fit.

## (b)

In [9]:
model5 = smf.ols(formula='RogueSales ~ Unemployment + RogueQueries + CPIAll + CPIEnergy + MonthFactor', data=rogue_train).fit()

print(model5.summary())

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.851
Model:                            OLS   Adj. R-squared:                  0.823
Method:                 Least Squares   F-statistic:                     30.46
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           7.49e-27
Time:                        12:52:31   Log-Likelihood:                -883.30
No. Observations:                  96   AIC:                             1799.
Df Residuals:                      80   BIC:                             1840.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

### i)
Describe your new model. Along with "Unemployment","CPIAll","CPIEnergy" and "RogueQueries", this new model has a new independent variable MonthFactror to take seasonality into consideration.
If we denote these variable "Unemployment" to $x_1$, "RogueQueries" to $x_2$, "CPIAll" to $x_3$, "CPIEnergy" to $x_4$, and the "MonthFactor" to $x_5$, the "RogueSales" to y, the regression equation is $$ y = {\beta_1}x_1 + {\beta_2}x_2 +{\beta_3}x_3 + {\beta_4}x_4 +{\beta_5}I(x_5)x_5 + {\alpha}$$
The ${\beta}$ is the corresponding coefficient in the above OLS regression result and ${\alpha}$ is the intercept.   

And ${\beta_1} = -266.84,  {\beta_2} = 111.26,  {\beta_3} = 486.46, {\beta_4} = -79.3$  

Note that there is an indicator function in the MonthFactor term. The MonthFactor is turned into a bunch of categorical variables, each of which will have data value 0 or 1 in each of the observation records.

Interpretation of MonthFactor: It represents the extra sales in a given month compared to the baseline month keeping other variables fixed.

### ii)
The R-Squared value for the new model is 0.851. And the significant variables are RogueQueries, MonthFactor[T.March], MonthFactor[T.August], MonthFactor[T.February],MonthFactor[T.May]

### iii)
Adding the independent variable MonthFactor does improve the quality of the model. First of all, we are able to discover more significant variables. Moreover, the R-Square value has increased compared to the previou one.

### iv)
I will choose to make a variable selection according to the p-value given in the previous regression analysis. If we are able to calculate with only significant independent monthly variable, we may improve the model since that many of the dummy variables are not statistically significant. Moreover, we can just change the data type from month to season, like spring\summer\autumn\winter to have a more specific or clear view of the change in terms of seasonality.

## (c)

In [10]:
model6 = smf.ols(formula='RogueSales ~ RogueQueries + CPIEnergy + MonthFactor', data=rogue_train).fit()

print(model6.summary())

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.777
Model:                            OLS   Adj. R-squared:                  0.742
Method:                 Least Squares   F-statistic:                     21.98
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           1.93e-21
Time:                        12:52:32   Log-Likelihood:                -902.66
No. Observations:                  96   AIC:                             1833.
Df Residuals:                      82   BIC:                             1869.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [11]:
def OSR2(model, df_train, df_test, dependent_var):   
    
    y_test = df_test[dependent_var]
    y_pred = model.predict(df_test)
    SSE = np.sum((y_test - y_pred)**2)
    SST = np.sum((y_test - np.mean(df_train[dependent_var]))**2)    
    
    return 1 - SSE/SST

In [12]:
OSR2(model6, rogue_train, rogue_testA, 'RogueSales')

0.9076908646686015

In [13]:
OSR2(model6, rogue_train, rogue_testB, 'RogueSales')

-0.31156177523204787

### i)
The variables I used in this model include "RogueQueries", "CPIEnergy" and "MonthFactor". The reason I choose "RogueQueries" and "MonthFactor" is same as the one I did in the previous, they are acting significant in the regression analysis.  
The "CPIEnergy", however, which was removed due to its not that statistical significance, it can be interpreted as if it increases, households have less purchasing power and there will be less car sales: if the price of gasoline increases for instance, less people will buy new cars. And as you can see that the coefficient of CPIenergy is negative.

### ii)
The training set R-Square value is 0.777

### iii)
The $OSR^2$ value for testing set A is 0.90769
The $OSR^2$ value for testing set B is -0.31156

### iv)
$ R^2$ is a coefficient of determination, measuring the overall quality of the regression model. The R-Square value here is 0.777, which is a comparably good fit.  
The $OSR^2$ s an assessment of the real-world performance of the model we have built. If OSR2is significantly smaller than R2(on the training data), this is an indicator of potential overfitting, however, it is not the case in our model.   
One thing that is very obvious is that the $ OSR^2 $ for the testB test set is negative. According to the formula of OSR, we can conclude that the Sum of squared residuals of regression model on the Test Set B is bigger than the Sum of squared residuals of baseline model applied to the Test Set B. We know that OSR>0 means that we improve compared to the baseline model, indicating the model is useful. If the OSR is negative, it may be due to the time series (the time of test set B is not as close to training set as test set A). And we have already known that the pandemic in 2020 and 2021 left a huge influencee on the RogueSales, which can not be anticipated just by the training set data. The negativity of OSR Squared value for test data B is reasonable and acceptable.

## (d)

In [14]:
ppi_data = pd.read_csv('ppi.csv')
rogue_new = pd.merge(rogue,ppi_data,on=['Year','MonthFactor'])

In [15]:
rogue_train1 = rogue_new[rogue_new['Year'] <= 2015]
rogue_testA1 = rogue_new[(rogue_new['Year'] >= 2016) & (rogue_new['Year'] <= 2019) ]
rogue_testB1 = rogue_new[(rogue_new['Year'] >= 2020) & (rogue_new['Year'] <= 2021) ]

In [16]:
model7 = smf.ols(formula='RogueSales ~ RogueQueries + CPIEnergy + MonthFactor + PPI', data=rogue_train1).fit()

print(model7.summary())

                            OLS Regression Results                            
Dep. Variable:             RogueSales   R-squared:                       0.778
Model:                            OLS   Adj. R-squared:                  0.739
Method:                 Least Squares   F-statistic:                     20.25
Date:                Tue, 21 Sep 2021   Prob (F-statistic):           8.13e-21
Time:                        12:52:35   Log-Likelihood:                -902.49
No. Observations:                  96   AIC:                             1835.
Df Residuals:                      81   BIC:                             1873.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [17]:
OSR2(model7, rogue_train1, rogue_testA1, 'RogueSales')

0.8988622063379132

In [18]:
OSR2(model7, rogue_train1, rogue_testB1, 'RogueSales')

-0.2666808850586413

This time I choose the variable PPI (a different economic indicator) The Producer Price Index (PPI) program measures the average change over time in the selling prices received by domestic producers for their output. The prices included in the PPI are from the first commercial transaction for many products and some services.

After conducting the regression analysis, I found that the R-Square has increased slightly from 0.777 to 0.778. As we all know, R-Square value will always increased by adding variables into the regression model. And moreover, the p-value of the new variable 'PPI' is 0.597, which is quite big, indicating that the new variable is not statistical significant.

We then move to the OSR-Squared value. We found that the OSR squared value decreased a bit.   

The new model does not add that much predictive value.

## (e)

According to out simplified assumption that the number of units carried over from month to month is always less than or equal to the target inventory, leading to the fact that the carried over inventory only lasted for one month for each unit.
Our lost function can be composed by three parts.  
I).   The scenario that the value of actual sales is larger than the value of expected sales. (The opportunity cost)  
II).  The scenario that the value of actual sales is smaller than the value of expected sales. (The carring over cost)  
III). At last, we need to take the profit into the loss funtion consideration.  

We use the indicator function to imply the relationship between the actual sales and expected sales.

The loss function is presented as followed:
$$ l(\hat{y},y) = 500\sum_{i=1}^{N}{(\hat{y}_i-y_i)}I(y_i<\hat{y}_i) + 3000\sum_{i=2}^{N}{(y_i-\hat{y}_i)}I(\hat{y}_i+(\hat{y}_{i-1}-y_{i-1},0)_+<y_i) - 3000\sum_{i=1}^{N}{\hat{y_i}}$$

where N is the total number of month.  

The first term in the loss function $500\sum_{i=1}^{N}{(\hat{y}_i-y_i)}I(y_i<\hat{y}_i)$ represents the scenario that the value of actual sales is smaller than the value of expected sales, such that the carrying over cost is occured. The cost per unit is 500. Moreover, according to our simplified assumption, the carried over inventory would not be carried again into the third following month. The carrying over cost is occured once for each unit.  

The second term of the loss function $3000\sum_{i=2}^{N}{(y_i-\hat{y}_i)}I(\hat{y}_i+(\hat{y}_{i-1}-y_{i-1},0)_+<y_i)$ represents the scenario that the value of actual value is larger than the value of expected sales, so that the opportunity cost of underestimating the sale is occured. The cost per unit is 3000.  
In this indicator function, we not only take the difference between actual and expected sales into consideration, we also need to pay attention to the fact that whether there will be carrying over inventory from the previous month to cover the opportunity cost. The term $(\hat{y}_{i-1}-y_{i-1},0)_+$ checks whether there will be carried-over inventory from last month. And if the expected sales plus the carried-over inventory is still less than the actual sales, we will face the opportunity cost.  

The last term is easy to understand. It is the total profit we made according to the expected sales schedule.